helusers/oidc.py

```diff
@@ -1,92 +1,4 @@
-from rest_framework.exceptions import AuthenticationFailed
-
-from .authz import UserAuthorization
-from .settings import api_token_auth_settings
-from .user_utils import get_or_create_user
-
-logger = logging.getLogger(__name__)
-
-
-class ApiTokenAuthentication(JSONWebTokenAuthentication):
-    def __init__(self, settings=None, **kwargs):
-        self.settings = settings or api_token_auth_settings
-        super(ApiTokenAuthentication, self).__init__(**kwargs)
-
-    @cached_property
-    def auth_scheme(self):
-        return self.settings.AUTH_SCHEME or 'Bearer'
-
-    @property
-    def oidc_config(self):
-        return self.get_oidc_config()
-
-    @cache(ttl=api_token_auth_settings.OIDC_CONFIG_EXPIRATION_TIME)
-    def get_oidc_config(self):
-        url = self.settings.ISSUER + '/.well-known/openid-configuration'
-        return requests.get(url).json()
-
-    def authenticate(self, request):
-        jwt_value = self.get_jwt_value(request)
-        if jwt_value is None:
-            return None
-
-        try:
-            payload = self.decode_jwt(jwt_value)
-        except AuthenticationFailed as e:
-            logger.debug("Invalid token signature")
-            raise
-
-        logger.debug("Token payload decoded as: {}".format(payload))
-
-        self.validate_claims(payload)
-
-        user_resolver = self.settings.USER_RESOLVER  # Default: resolve_user
-        user = user_resolver(request, payload)
-        auth = UserAuthorization(user, payload, self.settings)
-
-        if self.settings.REQUIRE_API_SCOPE_FOR_AUTHENTICATION:
-            api_scope = self.settings.API_SCOPE_PREFIX
-            if not auth.has_api_scope_with_prefix(api_scope):
-                raise AuthenticationFailed(
-                    _("Not authorized for API scope \"{api_scope}\"")
-                    .format(api_scope=api_scope))
-
-        return (user, auth)
-
-    def get_jwt_value(self, request):
-        auth = get_authorization_header(request).split()
-
-        logger.debug("Authorization header: {}".format(auth))
-
-        if not auth or smart_text(auth[0]).lower() != self.auth_scheme.lower():
-            return None
-
-        if len(auth) == 1:
-            raise AuthenticationFailed(
-                _("Invalid Authorization header. No credentials provided"))
-        elif len(auth) > 2:
-            raise AuthenticationFailed(
-                _("Invalid Authorization header. "
-                  "Credentials string should not contain spaces."))
-
-        return auth[1]
-
-    def authenticate_header(self, request):
-        return '{auth_scheme} realm="{realm}"'.format(
-            auth_scheme=self.auth_scheme,
-            realm=self.www_authenticate_realm)
-
-    def get_audiences(self, api_token):
-        return {self.settings.AUDIENCE}
-
-
-def resolve_user(request, payload):
-    return get_or_create_user(payload, oidc=True)
+try:
+    from ._oidc_auth_impl import ApiTokenAuthentication, resolve_user
+except ImportError:
```

Comment:
Is this imported in contexts where silent failure is needed?
I've been under the impression that ApiTokenAuthentication is imported only by users that will actually need the functionality.

Reply:
Not in this PR, but I did this so that the `helusers.oidc` module can be used for other OIDC related things as well. Those would be coming in future PRs.

Reply:
Is it important to preserve the old module name instead of just creating a new one? I do know it's a clean name... :)

Reply:
The name is so nice (as is `jwt`, which is also an existing module name), that I would really like to reuse these instead of inventing names that aren't the most appropriate.

Reply:
Approved.